## Analyzing Profitable Apps Profiles for Appstore and Google Play

by Erick G.

The goal of this project is to determine what kinds of apps would be the most profitable for apps whose only source of revenue would be in-app ads (free apps). Given that the kinds of apps that are the most popular would be the ones that would generate the most revenue, we want to focus our efforts on determining what apps would give the greatest financial return and focus on building these kinds of apps for the Appstore and Googe Play markets.

In [55]:

from csv import reader

### The Google Play data set ###
opened_file = open('googleplaystore.csv')
read_file = reader(opened_file)
android = list(read_file)
android_header = android[0]
android = android[1:]

### The App Store data set ###
opened_file = open('AppleStore.csv')
read_file = reader(opened_file)
ios = list(read_file)
ios_header = ios[0]
ios = ios[1:]


In [3]:
def explore_data(dataset, start, end, rows_and_columns=False):
    dataset_slice = dataset[start:end]    
    for row in dataset_slice:
        print(row)
        print('\n') # adds a new (empty) line between rows
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

print(android_header)
print('\n')
explore_data(android, 0, 3, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


['Coloring book moana', 'ART_AND_DESIGN', '3.9', '967', '14M', '500,000+', 'Free', '0', 'Everyone', 'Art & Design;Pretend Play', 'January 15, 2018', '2.0.0', '4.0.3 and up']


['U Launcher Lite – FREE Live Cool Themes, Hide Apps', 'ART_AND_DESIGN', '4.7', '87510', '8.7M', '5,000,000+', 'Free', '0', 'Everyone', 'Art & Design', 'August 1, 2018', '1.2.4', '4.0.3 and up']


Number of rows: 10841
Number of columns: 13


In [4]:
print(ios_header)
print('\n')
explore_data(ios, 0, 3, True)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


['389801252', 'Instagram', '113954816', 'USD', '0.0', '2161558', '1289', '4.5', '4.0', '10.23', '12+', 'Photo & Video', '37', '0', '29', '1']


['529479190', 'Clash of Clans', '116476928', 'USD', '0.0', '2130805', '579', '4.5', '4.5', '9.24.12', '9+', 'Games', '38', '5', '18', '1']


Number of rows: 7197
Number of columns: 16


It turns out that one of our entries is missing a value for 'content_rating'. We can see that it is blank (8th index value simply shows: ''

In [5]:
# DON'T RUN THIS
#print(android[10472])

Here, we'll delete it:

In [6]:
# DON'T RUN THIS
#del android[10472]

It is said that data analysts spend up to 80% of their time cleaning data, and the rest of their time analyzing it. For now, we must continue to clean the data so that our results are as accurate as possible. 

At times it happens that there are multiple entries for one app. Here are some examples from our Android dataset:

In [7]:
for app in android:
    name = app[0]
    if name == 'Instagram':
        print(app)

['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577446', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66577313', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']
['Instagram', 'SOCIAL', '4.5', '66509917', 'Varies with device', '1,000,000,000+', 'Free', '0', 'Teen', 'Social', 'July 31, 2018', 'Varies with device', 'Varies with device']


Let's now see how many duplicate entries there are in total:

In [8]:
duplicate_entries = []
unique_entries = []

for row in android:
    name = row[0]
    if name not in unique_entries:
        unique_entries.append(name)
    else:
        duplicate_entries.append(name)
print('All entries:', len(android))
print('Duplicate Entries:', len(duplicate_entries))
print('Unique entries:', len(unique_entries))

All entries: 10841
Duplicate Entries: 1181
Unique entries: 9660


We have 1,181 duplicate entries, but we will not delete them at random. We need to keep the entries with the most recent review count which is the one with the most reviews.

We must sift through the *duplicate_entries* list and isolate the entry with the most reviews.

In [9]:
most_reviews = {}

for row in android:
    name = row[0]
    reviews = row[3]
    if name in most_reviews and reviews > most_reviews[name]:
        most_reviews[name] = reviews
    if name not in most_reviews: # adds the entry to the dictionary for the first time
        most_reviews[name] = reviews
    
    

Now, let's take a peek at our unique apps with the most reviews:

In [10]:
print(most_reviews)
len(most_reviews)

{'CK Eorzea Timepiece': '8', 'Florida Cooling Supply HVAC': '6', 'Banana Kong': '3452530', 'Urban Limo Taxi Simulator': '824', 'Picai - Smart AI Camera': '529', 'Speech Therapy: CH': '0', 'Devise Dz': '26', 'GolfLogix GPS + Putt Breaks': '11085', 'AH of Rocky Hill': '0', 'Note Everything': '57033', 'FREE Stock Market Trading Tips': '714', 'Diabetes Testing': '1', 'Creative Destruction': '201426', 'Speed Car CM Locker Theme': '7519', 'EU VAT Checker': '7', 'Google Duo - High Quality Video Calls': '2083237', 'ei Calc': '2', 'CJ Auto école': '0', 'TUTORIAL CT SCAN CARDIAC': '1', 'Amazon FreeTime – Kids’ Videos, Books, & TV shows': '2715', 'P XPERIA Theme™ | INDIGO - Design For SONY 🎨': '12', 'Electricity Bill Calculator BD': '247', 'Jetting for Honda CR dirt bike': '7', 'Photo Compress 2.0 - Ad Free': '6267', 'FirstCry Baby & Kids Shopping, Fashion & Parenting': '41074', 'EZ Pass': '15', 'AS - News and sports results.': '20879', 'JetBlue': '24281', 'Coloring Book for Me & Mandala': '40121

9660

We can now use this information to remove the outdated entries--the ones with fewer reviews.


We will iterate through each row in the original Android dataset and compare the review count for each application, checking to see if it is the one with the most reviews by comparing it to what we have in our *most_reviews* dictionary. 

If we encounter a row with the highest amount of reviews, we will take the entire row and append it to our cleaned dataset; *android_clean*. 



In [11]:
android_clean = []
already_added = []

for row in android:
    name = row[0]
    reviews = row[3]
    if reviews == most_reviews[name] and name not in already_added:
        android_clean.append(row)
        already_added.append(name)
        
    

In [12]:
print(len(android_clean), 'is the amount of unique apps we now have in our dataset.')

9660 is the amount of unique apps we now have in our dataset.


Now, at this fictitious company, we are only interested in apps relevant to our market. This would exclude apps made in any language other than English. We will  need to remove them from our dataset. 

A simple way to do this is to look for the existence of any character that is not part of the alphabet in the title of the apps. To reduce the likelihood of false positives, we may later allow at least one character that falls outside of this range to allow for emojis, for example. For now, we will keep it as simple as possible to ensure that it works.

We use the built-in ord() function to return the number associated with each letter in the alphabet, which we iterate through by using our list.


In [13]:
alphabet_string = 'a b c d e f g h i j k l m n o p q r s t u v w x y z'

The only reason for the above code is that I thought it easier to quickly type the alphabet with spaces and ask Python to create the list for me:

In [14]:
alphabet_list = alphabet_string.split()

In [15]:
print(alphabet_list)

['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']


In [16]:
for letter in alphabet_list:
    print(letter,'corresponding number:', ord(letter))

a corresponding number: 97
b corresponding number: 98
c corresponding number: 99
d corresponding number: 100
e corresponding number: 101
f corresponding number: 102
g corresponding number: 103
h corresponding number: 104
i corresponding number: 105
j corresponding number: 106
k corresponding number: 107
l corresponding number: 108
m corresponding number: 109
n corresponding number: 110
o corresponding number: 111
p corresponding number: 112
q corresponding number: 113
r corresponding number: 114
s corresponding number: 115
t corresponding number: 116
u corresponding number: 117
v corresponding number: 118
w corresponding number: 119
x corresponding number: 120
y corresponding number: 121
z corresponding number: 122


But of course we can't really only depend on letters to determine if an app is made for the Anglophone market. Apparently, in addition to 97-122 being used, we also use the characters associated with 0-96. So for our purposes, if any character has an ordinal value greater than 127, we will toss it out.

In [17]:
print(chr(2))




We will now create a function to test if we can sort our app names appropriately.

We will begin by testing the following to see if they are detected as English or non-English -- that is to say, if they have individual characters whose ordinal number is greater than 127: 

* 'Instagram'
* '爱奇艺PPS -《欢乐颂2》电视剧热播'
* 'Docs To Go™ Free Office Suite'
* 'Instachat 😜'


Let's write our function...

In [18]:

def anglophoneAppChecker(app):
    for character in app:
        if ord(character) > 127:
            print('Foreign character', character, 'detected.')
            return False
    print('No foreign characters detected')
    return True
            

The above function will scan each letter of the app name, and should it find a character whose number is out of range (greater than 127), it will let us know that a foreign character was detected and return _False_. Otherwise, it will return _True_. 

This will be easy to interpret since the name of the function is called _anglophoneAppChecker()_ 

Let's test it out.

In [19]:
anglophoneAppChecker('Instagram')

No foreign characters detected


True

So far, so good. The function worked as it was supposed to.

In [20]:
anglophoneAppChecker('爱奇艺PPS -《欢乐颂2》电视剧热播')

Foreign character 爱 detected.


False

Let's try the next one on our list:

In [21]:
anglophoneAppChecker('Docs To Go™ Free Office Suite')

Foreign character ™ detected.


False

In [22]:
anglophoneAppChecker('Instachat 😜')

Foreign character 😜 detected.


False

Our simple English-app-detector seems to be working. However, had we used this to discard apps not written in English, we would have discarded _Docs To Go™ Free Office Suite_ and _Instachat 😜_ simply because these characters are associated with numbers greater than 127:

In [23]:
print(ord('™'))

8482


In [24]:
print(ord('😜'))

128540


We will edit the function to make it more useful to us. We will allow up to three foreign characters that fall ouside the range of 0-127. If we find more than three characters outside of this range, the function will return _False_; otherwise, it will return _True_.

In [25]:

def anglophoneAppChecker2(app):
    foreign_characters = 0 # to count each instance in each app name.
    for character in app:
        if ord(character) > 127 and foreign_characters < 3:
            foreign_characters += 1
            print(foreign_characters, 'foreign character', character, 'found')
        elif ord(character) < 127 and foreign_characters < 3:
            pass
        else:
            print("Scanning stopped. More than three foreign characters detected.")
            return False
            
    print('Three or fewer foreign characters detected.')
    return True
        

            

In [26]:
anglophoneAppChecker2('Instachat 😜')

1 foreign character 😜 found
Three or fewer foreign characters detected.


True

In [27]:
anglophoneAppChecker2('Docs To Go™ Free Office Suite')

1 foreign character ™ found
Three or fewer foreign characters detected.


True

In [28]:
anglophoneAppChecker2('爱奇艺PPS -《欢乐颂2》电视剧热播')

1 foreign character 爱 found
2 foreign character 奇 found
3 foreign character 艺 found
Scanning stopped. More than three foreign characters detected.


False

Just to be sure that it really works, I'll test three foreign characters to ensure that it returns _True_:

In [29]:
anglophoneAppChecker2('爱艺艺')

1 foreign character 爱 found
2 foreign character 艺 found
3 foreign character 艺 found
Three or fewer foreign characters detected.


True

I'll add another foreign character to ensure that it returns _False_:

In [30]:
anglophoneAppChecker2('爱艺艺爱')

1 foreign character 爱 found
2 foreign character 艺 found
3 foreign character 艺 found
Scanning stopped. More than three foreign characters detected.


False

Now let's remove the unecessary helper text and try this on our actual dataset: *android_clean*

Earlier, we created the android_clean dataset which removed the duplicate entries. Now we'll create a simple function to check if an app's name has more than three characters outside of the acceptable 'ASCII' range (i.e., greater than 127).

In [31]:
def app_in_English(string):
    foreign_characters = 0
    
    for character in string:
        if ord(character) > 127:
            foreign_characters +=1
    
    if foreign_characters > 3:
        return False
        
    else:
        return True
            
     

In [32]:
print(app_in_English('Instachat 😜'))

True


In [33]:
print(app_in_English('爱奇艺PPS -《欢乐颂2》电视剧热播'))

False


Great. It works.

Now let's use our new function on the *android_clean* dataset.

Earlier we saw that we had 9,659 apps. Let's see how many we have in our new, clean list; *android_clean_English*:

In [34]:
android_english = []
android_non_english = []

for app in android_clean:
    name = app[0]
    if app_in_English(name):
        android_english.append(name)
    else:
        android_non_english.append(name)
 
        

In [35]:
len(android_english)

9615

In [36]:
# apps removed

print(len(android_clean))

9660


Now, because we are only interested in free apps, we need to isolate those as well.

In [37]:
android_english_and_free = []

for app in android_clean:
    price = app[7]
    if price == '0':
        android_english_and_free.append(app)

In [38]:
len(android_english_and_free)

8903

Great. Now that we know it works, we can do the same thing for the iOS dataset. First, let's try to remember what the headers for the iOS dataset:

In [39]:
index_count = 0
print('HEADERS FOR IOS DATASET')
for column in ios_header:
    print(column, 'is at index', index_count)
    index_count+=1

HEADERS FOR IOS DATASET
id is at index 0
track_name is at index 1
size_bytes is at index 2
currency is at index 3
price is at index 4
rating_count_tot is at index 5
rating_count_ver is at index 6
user_rating is at index 7
user_rating_ver is at index 8
ver is at index 9
cont_rating is at index 10
prime_genre is at index 11
sup_devices.num is at index 12
ipadSc_urls.num is at index 13
lang.num is at index 14
vpp_lic is at index 15


In [40]:
duplicate_entries_ios = []
unique_entries_ios = []

for row in ios:
    name = row[1]
    if name not in unique_entries:
        unique_entries_ios.append(name)
    else:
        duplicate_entries_ios.append(name)
print('All entries:', len(ios))
print('Duplicate Entries:', len(duplicate_entries_ios))
print('Unique entries:', len(unique_entries_ios))

All entries: 7197
Duplicate Entries: 328
Unique entries: 6869


In [41]:
most_reviews_ios = {}

for row in ios:
    name = row[1]
    reviews = row[5]
    if name in most_reviews_ios and reviews > most_reviews_ios[name]:
        most_reviews_ios[name] = reviews
    if name not in most_reviews_ios: # adds the entry to the dictionary for the first time
        most_reviews_ios[name] = reviews
    

In [42]:
ios_clean = []
ios_already_added = []

for row in ios:
    name = row[1]
    reviews = row[5]
    if reviews == most_reviews_ios[name] and name not in ios_already_added:
        ios_clean.append(row)
        ios_already_added.append(name)
        

In [43]:
print(len(ios_clean))

7195


In [44]:
ios_english = []
ios_non_english = []

for app in ios_clean:
    name = app[1]
    if app_in_English(name):
        ios_english.append(name)
    else:
        ios_non_english.append(name)

In [45]:
print(len(ios_english))

6181


Now, we must run our little function to find the free applications.

In [46]:
ios_english_and_free = []

for app in ios_clean:
    price = app[4]
    if price == '0.0':
        ios_english_and_free.append(app)

##### Finally, we now have the final version of our datasets that we want; unique, free, and in English.

In [47]:
print('Length android dataset: ', len(android_english_and_free), '\n',
     'Length ios dataset: ', len(ios_english_and_free))

Length android dataset:  8903 
 Length ios dataset:  4054


## So far, we have done the following:
* removed duplicate entries, keeping only the most recently updated ones
* removed non-English apps
* removed the non-free apps

Still, we aren't finished. Since our main source of revenue will be in-app advertising, we need to maximize our exposure and therefore find the apps that are the most popular. 

"To minimize risks and overhead, our validation strategy for an app idea is comprised of three steps:

Build a minimal Android version of the app, and add it to Google Play.
If the app has a good response from users, we develop it further.
If the app is profitable after six months, we build an iOS version of the app and add it to the App Store."

We eventually want to be on both the Google Play and App Store, so we need to find "app profiles that are successful on both markets."

So now we will begin analyzing our datasets to get a better sense of the most common genres for each market. A frequency table may help us better visualize this data.

Since we're interested in the genres of each dataset, we could begin by seeing which ones are the most frequent using...frequency tables!

Let's start by trying to remember what column our genres are in for both datasets.

In [48]:
print('Android header:',android_header)


Android header: ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


In [49]:
print('iOS header:', ios_header)

iOS header: ['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


The Android genre is in index 9 and is named 'Genres'

The iOS genre is in index 11 and is named 'prime_genre'

We have all the information we need to create histograms for both datasets.

In [50]:
# Counts the instances of categories, for example:

def freq_table(dataset, index):
    table = {}
    total = 0
    
    for row in dataset:
        total += 1 #counts the rows
        value = row[index]
        if value in table:
            table[value] += 1
        else:
            table[value] = 1
    #Using the tally now in the 'table' dictionary, we calculate the
    # percentages
    
    table_percentages = {}
    for key in table: # for each entry in the table dict
        percentage = (table[key] / total) * 100 #define percentage
        table_percentages[key] = percentage 
    
    return table_percentages


def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)
        
    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

We want to find an app profile that works for both the App Store and Google Play. 

In [51]:
display_table(ios_english_and_free, -5)

Games : 55.6240749876665
Entertainment : 8.238776517020227
Photo & Video : 4.119388258510114
Social Networking : 3.5273803650715343
Education : 3.2560434139121854
Shopping : 2.9847064627528366
Utilities : 2.688702516033547
Lifestyle : 2.318697582634435
Finance : 2.0720276270350273
Sports : 1.9486926492353234
Health & Fitness : 1.8746916625555006
Music : 1.6526887025160337
Book : 1.6280217069560927
Productivity : 1.5293537247163296
News : 1.4306857424765662
Travel : 1.3813517513566849
Food & Drink : 1.0606808090774542
Weather : 0.7646768623581648
Reference : 0.493339911198816
Navigation : 0.493339911198816
Business : 0.493339911198816
Catalogs : 0.2220029600394672
Medical : 0.19733596447952642


We can see that among the English apps in the App Store, the majority are games. The top four genres actually are a bit more granular than they need to be. For example, games could be a subcategory of entertainment. Either way, the most popular apps have some sort of social component (playing games with friends, or social networking, or apps used for photography which can presumably be shared on social networks.

Interestingly enough, education is right below social networking, however we must be careful to not give too much attention to apps that are not used often or do not have much engagement. There is nothing in our data that represents the amount of time spent on the apps. 

Let's also take a look at the Google Play store:

In [52]:
display_table(android_english_and_free, 1)

FAMILY : 19.00482983264068
GAME : 9.670897450297652
TOOLS : 8.424126698865551
BUSINESS : 4.58272492418286
LIFESTYLE : 3.931259126137257
PRODUCTIVITY : 3.8863304504099743
FINANCE : 3.6841514096372006
MEDICAL : 3.5044367067280695
SPORTS : 3.380882848478041
PERSONALIZATION : 3.3134898348871165
COMMUNICATION : 3.2348646523643714
HEALTH_AND_FITNESS : 3.0663821183870605
PHOTOGRAPHY : 2.9428282601370324
NEWS_AND_MAGAZINES : 2.8305065708188253
SOCIAL : 2.6507918679096933
TRAVEL_AND_LOCAL : 2.3250589688868923
SHOPPING : 2.246433786364147
BOOKS_AND_REFERENCE : 2.179040772773223
DATING : 1.8533078737504212
VIDEO_PLAYERS : 1.7971470290913176
MAPS_AND_NAVIGATION : 1.4152532854094126
FOOD_AND_DRINK : 1.2355385825002807
EDUCATION : 1.1793777378411772
ENTERTAINMENT : 0.9547343592047625
LIBRARIES_AND_DEMO : 0.932270021341121
AUTO_AND_VEHICLES : 0.9210378524093001
HOUSE_AND_HOME : 0.8199483320229137
WEATHER : 0.7974839941592722
EVENTS : 0.7076266427047063
PARENTING : 0.6514657980456027
ART_AND_DESIGN : 

So far, we still see games at the very top of the list. It appears that Android users don't use their phones for games as much as the iPhone users. That could be because maybe iPhone users tend to be younger in general which may mean more time for leisure. Either way, games seems to be a safe choice as it makes the top of the list on both profiles.

The frequency tables we have analyzed so far shows us that Apple's Play Store is dominated by apps designed for fun, which the Google Play store shows a more balanced landscape. 

However, the most apps do not necessarily mean that they have the most users. So, let's try to get an idea about the kinds of apps that are used the most.

One strategy for determining this is to calculate the number of installations for each genre. 

For the Google Play store, this is found at index 5; the 'Installs' column.

In [57]:
print(android_header)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


However, the App Store doesn't have this:

In [58]:
print(ios_header)

['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


What we'll use then is the amount of ratings, using 'rating_count_tot', at index 5. If the app was rated, it was likely downloaded. This isn't perfect, but it's the next best option we have.

Let's start with calculating the average number of user ratings per app genre on the App Store. 

In [59]:
#Find the average amount of ratings per genre.
#This will be done by adding up all the ratings of each genre,
#then dividing that by amount of apps in that genre.

In [60]:
#First, let's find the amount of ratings for each genre:

In [72]:
counter_dict = {}


def instance_counter(dataset, tally_index, mykeyword, index_to_search_for_my_keyword): 
    tally = dataset[tally_index]
    foundkeyword = dataset[index_to_search_for_my_keyword]
    for row in dataset:
        if mykeyword == foundkeyword and foundkeyword in counter_dict:
            counter_dict[foundkeyword] += tally
        elif mykeyword == foundkeyword and foundkeyword not in counter_dict:
            counter_dict[foundkeyword] = tally

    return counter_dict
        
        
        
        

In [73]:
instance_counter(ios_english_and_free, 5, 'rating_count_tot', -5)

{}